[clang-tidy] false positive on bugprone-assignment-in-if-condition with a requires clause in if

Tested on 902fb1b4653d5a23613492406cd5693446f06ab6

# Steps to reproduce:
test.cpp
```c++
int main() {
	if constexpr (requires (int &a) { a = 0; }) {}
}
```
```sh
clang-tidy --checks=bugprone-assignment-in-if-condition ./test.cpp -- -std=c++20
```

## result:
```
1 warning generated.
test.cpp:2:38: warning: an assignment within an 'if' condition is bug-prone [bugprone-assignment-in-if-condition]
    2 |         if constexpr (requires (int &a) { a = 3; }) {}
      |                                           ~~^~~
test.cpp:2:38: note: if it should be an assignment, move it out of the 'if' condition
test.cpp:2:38: note: if it is meant to be an equality check, change '=' to '=='
```

--- a/clang-tools-extra/clang-tidy/bugprone/AssignmentInIfConditionCheck.cpp
+++ b/clang-tools-extra/clang-tidy/bugprone/AssignmentInIfConditionCheck.cpp
@@ -39,6 +39,12 @@ void AssignmentInIfConditionCheck::check(
           return true;
         }
 
+        // Dont traverse into any requires expressions.
+        bool TraverseRequiresExpr(RequiresExpr *,
+                                  DataRecursionQueue * = nullptr) {
+          return true;
+        }
+
         bool VisitBinaryOperator(BinaryOperator *BO) {
           if (BO->isAssignmentOp())
             Check.report(BO);

--- a/clang-tools-extra/docs/ReleaseNotes.rst
+++ b/clang-tools-extra/docs/ReleaseNotes.rst
@@ -232,6 +232,10 @@ Changes in existing checks
   <clang-tidy/checks/bugprone/assert-side-effect>` check by detecting side
   effect from calling a method with non-const reference parameters.
 
+- Improved :doc:`bugprone-assignment-in-if-condition
+  <clang-tidy/checks/bugprone/assignment-in-if-condition>` check by ignoring
+  assignments in the C++20 ``requires`` clause.
+
 - Improved :doc:`bugprone-casting-through-void
   <clang-tidy/checks/bugprone/casting-through-void>` check by ignoring casts
   where source is already a ``void``` pointer, making middle ``void`` pointer

--- a/clang-tools-extra/test/clang-tidy/checkers/bugprone/assignment-in-if-condition-cxx20.cpp
+++ b/clang-tools-extra/test/clang-tidy/checkers/bugprone/assignment-in-if-condition-cxx20.cpp
@@ -1,0 +1,6 @@
+// RUN: %check_clang_tidy -std=c++20 %s bugprone-assignment-in-if-condition %t
+
+void testRequires() {
+  if constexpr (requires(int &a) { a = 0; }) {
+  }
+}

